types/externs.d.ts

```diff
@@ -29,8 +53,16 @@ declare global {
   /** Make optional all properties on T and any properties on object properties of T. */
   type RecursivePartial<T> = {
     [P in keyof T]+?:
+      // If type is a union, map each individual component and transform the resultant tuple back into a union.
+      // Only up to 4 components is supported. For more, modify the following line and `IntersectionOfFunctionsToType`.
+      // Guard against large string unions, which would be unreasonable to support (much more than 3 components is common).
+      SplitType<T[P]> extends string[] ? T[P] :
```

Comment:
wouldn't this be guarded against already by `GetLength<SplitType<T[P]>>` being `never`?

Reply:
what do you mean?

![image](https://user-images.githubusercontent.com/4071474/72302672-b104f680-361f-11ea-844c-bdbe47f4cf87.png)

note that one component was lost, due to the 4 component constraint.

Reply:
oh so `IntersectionOfFunctionsToType` silently drops it, the lack of support doesn't mean fall into the `never` branch? I guess it is `F extends`  not `F is`....

alright alright I think I get it now

does `Only the first 4 components will be captured and the rest will be dropped.` work for you instead of `Only up to 4 components is supported`? :)

Reply:
> does Only the first 4 components will be captured and the rest will be dropped. work for you instead of Only up to 4 components is supported? :)

Hmm, not exactly.

nit: /first four/last four/s.

I think how the unsupported behavior manifests is less important than that it's unsupported :)

typescript will fail horribly if type `T` is applied to type `RecursivePartial<T>`, if `T` contains some (non-string) union with >4 components. so saying "just the last 4 components of a union will exist for a property" doesn't matter much to the caller - the type will fail, and the only way to fix it is amend these utility types.

Reply:
> I think how the unsupported behavior manifests is less important than that it's unsupported :)

1. `Only the first 4 components will be captured and the rest will be dropped.` is super obvious to me personally that it's unsupported so I didn't see this as the loss of any information or emphasis. Happy to reword however you see fit to not lose that unsupported bit.
1. Importance to the code overall, sure. Importance to the person who is reading this comment, I'm less convinced :) The only reason you'd be reading this specific code at all is to figure out what these types do, the surrounding code has a pretty clear manual copy-paste to support n-things pattern that is somewhat common in other typescript use cases that a reader could pickup on while the specific failure mode for this usage of the pattern is rather unique IMO that led to a misunderstanding and conversation already, so I don't think the harm by including this information in a comment is sufficiently detrimental to avoid it.

> so saying "just the last 4 components of a union will exist for a property" doesn't matter much to the caller

Agreed it doesn't matter much to the caller, but it matters to the person trying to comprehend the complex type expression here and callers aren't reading this comment :)

Reply:
how's what I just pushed?

Reply:
looks great!